staging/src/k8s.io/apiserver/pkg/storage/names/generate.go

```diff
@@ -42,7 +42,7 @@ var SimpleNameGenerator NameGenerator = simpleNameGenerator{}
 const (
 	// TODO: make this flexible for non-core resources with alternate naming rules.
 	maxNameLength          = 63
-	randomLength           = 5
+	randomLength           = 10
```

Comment:
cc @alculquicondor in case he can think of any other places we had controllers doing fragile things with generated names

all the places I can find are referencing MaxGeneratedNameLength, which will ~work for in-tree version-paired things... anything using that constant with version skew now will be disappointed

Reply:
example at https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/controller/store/endpointslice.go#L44-L52 ... I didn't reason through whether that code would fail safe or not if the server had a different view of MaxGeneratedNameLength

Reply:
Should we nuke that constant?

Reply:
probably? I have a hard time imagining it being used for anything good

Reply:
Without a constant, job-controller will have a hard time inserting its oridinal into the name.  Still, even that is suspect.  Names are not for parsing.  Names are not for parsing.  Names are not for parsing.

Reply:
Maybe we should revisit one of the alternative ideas - make `generateName` optionally specify the number of random characters, a la mktemp.

E.g. "base" is equivalent to "base-XXXXX", but users or controllers could say "base-XXXXXXXXXX".  Job-controller can now own how many characters it uses, obviating the need for a constant.

We can still change the default, or not, as an independent decision. Or each controller can, on their own, control how they adopt this idea.

Could be "XXXXX" like mktemp or "*****" or something else.

Reply:
That's interesting... controllers would need to wait until they were sure they were talking to a new enough API server to start using that (and a controller talking through a new kube-apiserver to an old aggregated API server is an interesting edge), but all existing `generateName` users would be unimpacted, depending on the character we chose to represent the generated suffix.

A character we know is not allowed in object names would be ideal... % comes to mind since it is globally forbidden by generic ObjectMeta validation.

Would we allow *fewer* than 5 random characters, letting a controller *increase* the likelihood of conflicts beyond the current state? Probably not.

Reply:
Yes, the job controller would fail safe: the name would just loose the index. But that's why the index is part of an annotation. Job names are actually validated to accomodate `-{max_completions}`, so that the hostname is still `job-name-{index}`

Reply:
> A character we know is not allowed in object names would be ideal... % comes to mind since it is globally forbidden by generic ObjectMeta validation.

Where is that?  I don't see it.

> Would we allow fewer than 5 random characters, letting a controller increase the likelihood of conflicts beyond the current state? Probably not.

Agree not.

Reply:
> Where is that? I don't see it.

ValidatePathSegmentName, called on all objects in `BeforeCreate`

https://github.com/kubernetes/kubernetes/blob/d3a7b5920fc4a443c6e48c98478d1eecf8478246/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L129-L131

Reply:
Ahh yes.

```
var NameMayNotContain = []string{"/", "%"}
```

That should probably include '*', '&', '?', ',', ';', and more.  We can audit all builtin types, but we can't say for sure if CRDs or aggregated types use any of those.  What's your thought on risk assessment here?

Reply:
CRD names are limited to RFC 11123

Reply:
I wouldn't expand what that function forbids at all. It was solely to ensure every name we accept is safe as a URL path segment. Forbidding % was done because different servers/clients treat %-encoding differently, especially around `%2F`

As Aldo noted, custom resource names are already limited to RFC 1123 names.

Reply:
But aggregated APIservers are not so restricted, so even adding '*' to the NameMayNotContain list is probably a non-starter, though I cant imagine it being something people ACTUALLY do.  Sigh, Hyrum's law sucks.

So % is the only really viable option if we pursue this opt-in approach.

Or, and I know @lavalamp isn't super keen on this, we fix the apiserver to retry when random generation results in a collision.